validator.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * see https://github.com/dequelabs/axe-core/blob/develop/doc/rule-descriptions.md
+ *
+ * must have input element
+ *
+ * if element provided with *-clear, must be button element
+ *
+ * 'label-title-only'
+ *  must have 1 of:
+ * - label
+ * - aria-label
+ * - aria-labelledby
+ */
+
+const generatedRules = []
+const generatedChecks = []
+function createRule({id, excludeHidden = true, selector, checks = [], all = [], any, help, helpUrl = ''}) {
+  if (/[^,] /.test(selector)) {
+    // eslint-disable-next-line no-console
+    console.warn('Try to avoid nested CSS selectors in `createRule`: %o', selector)
+  }
+  if (checks) {
+    checks.map((func, i) => {
+      const checkId = `${id}_${i}`
+      if (all) all.push(checkId)
+      generatedChecks.push({
+        id: checkId,
+        evaluate: func,
+        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+        // @ts-ignore FIXME: axe-core does not list this type. Is it part of the API?
+        metadata: {impact: 'critical'}
+      })
+    })
+  }
+  /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+  // @ts-ignore FIXME: axe-core does not list metadata in part of its type. Is it part of the API?
+  generatedRules.push({id, excludeHidden, selector, all, any, metadata: {help, helpUrl}})
+}
+
+function checkForInput(htmlString) {
+  let container
+  if (htmlString.ownerDocument) {
+    // axe tests
+    container = htmlString.ownerDocument.getElementById('mocha-fixture')
+  } else {
+    // javascript tests
+    container = document.getElementById('mocha-fixture')
+    container.innerHTML = htmlString
+  }
+
+  return !!container.getElementsByTagName('INPUT').length
+}
+function checkForOptionalClearButton(htmlString) {
+  let container
+  if (htmlString.ownerDocument) {
+    // axe tests
+    container = htmlString.ownerDocument.getElementById('mocha-fixture')
+  } else {
+    // javascript tests
+    container = document.getElementById('mocha-fixture')
+    container.innerHTML = htmlString
+  }
+
+  const [input] = container.getElementsByTagName('INPUT')
+  const clearButtonId = `${input.id || input.getAttribute('name')}-clear`
+  const clearButton = document.getElementById(clearButtonId)
+  if (clearButton && !(clearButton instanceof HTMLButtonElement)) {
+    return false
+  }
+  return true
+}
+createRule({
+  id: 'for-input',
+  // excludeHidden: false,
+  selector: 'auto-complete',
+  help: 'To know about input you should know X',
+  checks: [checkForInput]
+})
+createRule({
+  id: 'for-clear-button',
+  // excludeHidden: false,
+  selector: 'auto-complete',
+  help: 'To know about clear button you should know Y',
+  checks: [checkForOptionalClearButton]
```

Comment:
ids and help text should be simple and if relevant point to Deque documentation explaining why, I think. curious to hear suggestions from others.